### ISSUE TITLE
BigGAN LR: auto-tune learning rate using gradient noise scale?

### ISSUE THREAD

**Issue by gwern** [tensorfork/tensorfork#33]
The OA [GPT-3](https://arxiv.org/pdf/2005.14165.pdf#page=9) paper notes:

> As found in [KMH+20,MKAT18], larger models can typically use a larger batch size, but require a smaller learning rate. We measure the gradient noise scale during training and use it to guide our choice of batch size [MKAT18].

The 'gradient noise scale' was introduced in ["An Empirical Model of Large-Batch Training"](https://arxiv.org/abs/1812.06162), McCandlish et al 2018 ([blog](https://openai.com/blog/science-of-ai/ "How AI Training Scales: We’ve discovered that the gradient noise scale, a simple statistical metric, predicts the parallelizability of neural network training on a wide range of tasks. Since complex tasks tend to have noisier gradients, increasingly large batch sizes are likely to become useful in the future, removing one potential limit to further growth of AI systems. More broadly, these results show that neural network training need not be considered a mysterious art, but can be rigorized and systematized."), [tips](https://ufal.mff.cuni.cz/pbml/110/art-popel-bojar.pdf#page=21 "'Training Tips for the Transformer Model', Poel & Bojar 2018")):

> In an increasing number of domains it has been demonstrated that deep learning models can be trained using relatively large batch sizes without sacrificing data efficiency. However the limits of this massive data parallelism seem to differ from domain to domain, ranging from batches of tens of thousands in ImageNet to batches of millions in RL agents that play the game Dota 2. To our knowledge there is limited conceptual understanding of why these limits to batch size differ or how we might choose the correct batch size in a new domain. In this paper, we demonstrate that a simple and easy-to-measure statistic called the gradient noise scale predicts the largest useful batch size across many domains and applications, including a number of supervised learning datasets (MNIST, SVHN, CIFAR-10, ImageNet, Billion Word), reinforcement learning domains (Atari and Dota), and even generative model training (autoencoders on SVHN). We find that the noise scale increases as the loss decreases over a training run and depends on the model size primarily through improved model performance. Our empirically-motivated theory also describes the tradeoff between compute-efficiency and time-efficiency, and provides a rough model of the benefits of adaptive batch-size training.

The intuition is that as minibatch sizes increase, they increasingly closely approximate the true gradient, the ideal direction in which to tweak the model; if the minibatch size goes beyond that, the extra compute is largely wasted, while if the minibatch is too small, one could throw some more GPUs at the problem to gain a wallclock speedup. The ideal minibatch size will depend heavily on the intrinsic difficulty of the task and also how useful feedback the NN gets from the data/training-process. In a case like supervised learning using pixel-level segmentation on simple small images, such extremely clear noise-free feedback on the right answer may permit tiny minibatches _n_<100; on the other end of things, DoTA 2 is an extremely complex stochastic environment with long-range feedback and highly subtle tradeoffs where the best answer is unknown where the best algorithm (PPO) is quite dumb/blind & playing against itself, and so the gradient noise scale reveals that the optimal minibatch is _n_ ~ millions (more:  ["Dota 2 with Large Scale Deep Reinforcement Learning"](https://cdn.openai.com/dota-2.pdf), Berner et al 2019). Typically, the optimal minibatch size increases over the course of training, as the easy problems are solved and the NN must learn subtler aspects of the tasks; so if the minibatch size is left at a constant size, it either wastes compute or wallclock time (or requires reducing learning rate to keep too-small minibatches from over-updating the model based on garbage gradients).

As has been noted before, actor-critic/self-play DRL is closely connected to GANs, and of course, BigGAN benefits considerably from larger minibatches at least up to _n_=2048.

Does gradient noise scale work for GANs, particularly our very big GANs? It seems that no one has ever tried. It's possible that we could get BigGAN to iterate much faster if early on, the gradient noise scale indicates much smaller minibatches than our usual 2048 or 4096, and if the noise scale keeps pushing up the batch size later on in training, perhaps that would justify implementing gradient accumulation #16 sooner rather than later (and might also explain the difficulty of 'dialing in' on sharp details later in training and the need for [SWA](https://armenag.com/posts/2019-05-13/ " Stochastic Weight Averaging and the Ornstein-Uhlenbeck Process")/[EMA](https://arxiv.org/abs/1806.04498 "'The Unusual Effectiveness of Averaging in GAN Training', Yazici et al 2018") in StyleGAN/BigGAN).

An example of gradient noise scale in RLLib, although it didn't seem to quite work out for them judging by their bug reports: https://github.com/ray-project/ray/commit/7f8dd009c581c206b0819a6c585e1eb332a6b0fa ; a [KungFu TensorFlow library](https://www.imperial.ac.uk/media/imperial-college/faculty-of-engineering/computing/public/1819-ug-projects/BrabeteA-Kungfu-A-Novel-Distributed-Training-System-for-TensorFlow-using-Flexible-Synchronisation.pdf#page=86) has [an implementation here](https://github.com/lsds/KungFu/blob/master/srcs/python/kungfu/tensorflow/optimizers/grad_noise_scale.py); this [may or may not be](https://github.com/mldbai/tensorflow-models/blob/master/neural_gpu/neural_gpu_trainer.py#L41) another one.
